src/services/game-logic/GameEngine.ts

```diff
@@ -12,16 +12,141 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let pieceMoves = piece?.getPossibleMoves(this.board) ?? [];
-        return pieceMoves.filter(this.isOnBoard).filter((square) => !this.isOccupiedBySameColor(square, piece));
+        const pieceMoves = piece!.getPossibleMoves(this.board);
+
+        const firstCollission = this.getFirstCollission(square);
+        const pieceMoves2 = pieceMoves
+            .filter(this.isOnBoard)
+            .filter((square) => !this.isOccupiedBySameColor(square, piece));
+
+        if (piece?.name === PieceNames.ROOK || piece?.name === PieceNames.QUEEN) {
+            const rook = pieceMoves2
+                .filter((square) => !(square.row < firstCollission[0] && square.column === piece.position.column))
+                .filter((square) => !(square.row > firstCollission[1] && square.column === piece.position.column))
+                .filter((square) => !(square.column < firstCollission[2] && square.row === piece.position.row))
+                .filter((square) => !(square.column > firstCollission[3] && square.row === piece.position.row))
+                .filter((square) => !(square.row < firstCollission[4] && square.column < piece.position.column))
+                .filter((square) => !(square.row > firstCollission[5] && square.column < piece.position.column))
+                .filter((square) => !(square.row < firstCollission[6] && square.column > piece.position.column))
+                .filter((square) => !(square.row > firstCollission[7] && square.column > piece.position.column));
+            return rook;
+        } else if (piece?.name === PieceNames.BISHOP) {
+            const bishop = pieceMoves2
+                .filter((square) => !(square.row < firstCollission[4] && square.column < piece.position.column))
+                .filter((square) => !(square.row > firstCollission[5] && square.column < piece.position.column))
+                .filter((square) => !(square.row < firstCollission[6] && square.column > piece.position.column))
+                .filter((square) => !(square.row > firstCollission[7] && square.column > piece.position.column));
+
+            return bishop;
+        } else return pieceMoves2;
+    };
+    private detectAllCollisions = (square: Square) => {
+        const PossibleMoves = this.board.getPiece(square)?.getPossibleMoves(this.board).flat();
+
+        const piecesPositionsOnBoard = this.board.state
+            .map((e) => Object.entries(e).map(([, y]) => y?.position))
```

Comment:
jeśli nie potrzebujesz zmiennej to w JS widziałem stosuje się podłogę, czyli .map([_,y[) => ...

Reply:
oki, wrzuciłam z _